crates/ruff_graph/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+use crate::collector::Collector;
+pub use crate::db::ModuleDb;
+use crate::resolver::Resolver;
+pub use crate::settings::{Direction, GraphSettings};
+use anyhow::Result;
+use ruff_python_ast::helpers::to_module_path;
+use ruff_python_parser::{parse, Mode};
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+use std::path::{Path, PathBuf};
+
+mod collector;
+mod db;
+mod resolver;
+mod settings;
+
+#[derive(Debug, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct ModuleImports(BTreeSet<PathBuf>);
+
+impl ModuleImports {
+    pub fn insert(&mut self, path: PathBuf) {
+        self.0.insert(path);
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Convert the file paths to be relative to a given path.
+    #[must_use]
+    pub fn relative_to(&self, path: &Path) -> Self {
+        let mut imports = Self::default();
+        for import in &self.0 {
+            if let Ok(path) = import.strip_prefix(path) {
+                imports.insert(path.to_path_buf());
+            }
+        }
+        imports
+    }
+}
+
+#[derive(Debug, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct ImportMap(BTreeMap<PathBuf, ModuleImports>);
+
+impl ImportMap {
+    pub fn insert(&mut self, path: PathBuf, imports: ModuleImports) {
+        self.0.insert(path, imports);
+    }
+
+    #[must_use]
+    pub fn reverse(imports: impl IntoIterator<Item = (PathBuf, ModuleImports)>) -> Self {
+        let mut reverse = ImportMap::default();
+        for (path, imports) in imports {
+            for import in imports.0 {
+                reverse.0.entry(import).or_default().insert(path.clone());
+            }
+        }
+        reverse
+    }
+}
+
+impl FromIterator<(PathBuf, ModuleImports)> for ImportMap {
+    fn from_iter<I: IntoIterator<Item = (PathBuf, ModuleImports)>>(iter: I) -> Self {
+        let mut map = ImportMap::default();
+        for (path, imports) in iter {
+            map.0.entry(path).or_default().0.extend(imports.0);
+        }
+        map
+    }
+}
+
+/// Generate the module imports for a given Python file.
+pub fn generate(
+    path: &Path,
```

Comment:
I would recommend you to use `SystemPath` in more places and do the conversation from `Path` at the boundary rather than ad-hoc in here.